### ISSUE TITLE
model.fit时的验证可否直接用validation_data参数

### ISSUE THREAD

**Issue by canvas7** [bojone/bert4keras#356]
提问时请尽可能提供如下信息：

### 基本信息
- 你使用的**操作系统**:  Windows 10
- 你使用的**Python**版本:  3.8
- 你使用的**Tensorflow**版本:  2.2.2
- 你使用的**Keras**版本:  2.3.1
- 你使用的**bert4keras**版本:  0.10.6
- 你使用纯**keras**还是**tf.keras**:  tf.keras
- 你加载的**预训练模型**: albert

### 核心代码
```python
history = model.fit(train_generator.forfit(),
                    steps_per_epoch=len(train_generator),
                    epochs=epochs,
                    validation_data=valid_generator.forfit(),
                    callbacks=[early_callback, ckpt_callback],
                    verbose=1)
```

### 输出信息
```shell
一直卡在第一个epoch。
Epoch 1/10
150/150 [==============================] - ETA: 0s - loss: 0.2767 - accuracy: 0.9113 - f1_score_1: 0.0385
```

看到example里是用callback的形式写的验证和测试，想尝试直接用validation_data参数，但是不知道为什么一直卡住，看起来是valid没有计算成功，但是也没有报错，想知道是为什么？是否只能用callback的形式去计算？

**Comment by i4never** [comment]
哪个任务？example里的task脚本、evaluate没怎么优化（batch predict之类的）、个别任务下甚至可能出现耗时比train更长的情况、一个是可以加个tqdm反馈下evaluate的进度、然后嫌慢就自己优化评估的计算过程（主要是一批一批样本处理）、

**Comment by canvas7** [comment]
主要参考的是这个task_sentiment_albert.py: 任务例子，情感分类任务，加载ALBERT模型。我看基本上所有例子都是类似的。


    model.fit(
        train_generator.forfit(),
        steps_per_epoch=len(train_generator),
        epochs=20,
        callbacks=[evaluator]
    ) 


我的意思是源代码没问题，是用“callbacks=[evaluator]"实现的evaluate和predict。
我自己改写的代码有问题，用model.fit里的validation_data参数，也赋予一个generator，如‘validation_data=valid_generator.forfit()’，但是却一直卡住，不知道为什么。

------------------ 原始邮件 ------------------
发件人:                                                                                                                        "bojone/bert4keras"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年6月4日(星期五) 下午2:32
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [bojone/bert4keras] model.fit时的验证可否直接用validation_data参数 (#356)





 
哪个任务？example里的task脚本、evaluate没怎么优化（batch predict之类的）、个别任务下甚至可能出现耗时比train更长的情况、一个是可以加个tqdm反馈下evaluate的进度、然后嫌慢就自己优化评估的计算过程（主要是一批一批样本处理）、

**Comment by canvas7** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
    
        问题已解决，需要设置validation_steps=len(valid_generator)，否则会陷入forfit的无限循环中。
    
        
            
        
    
    
     
    
        
                    
            
                
                    
                            
                                
                            
                            
                                598756381
                            
                    
                        
                            
                                    ***@***.***
                            
                        
                
            
        
        
    
    
        签名由
        网易邮箱大师
        定制
    
 

    在2021年6月4日 ***@***.***> 写道： 



哪个任务？example里的task脚本、evaluate没怎么优化（batch predict之类的）、个别任务下甚至可能出现耗时比train更长的情况、一个是可以加个tqdm反馈下evaluate的进度、然后嫌慢就自己优化评估的计算过程（主要是一批一批样本处理）、

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by canvas7** [comment]
问题已解决，需要设置validation_steps=len(valid_generator)，否则会陷入forfit的无限循环中。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "bojone/bert4keras"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年6月4日(星期五) 下午2:32
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [bojone/bert4keras] model.fit时的验证可否直接用validation_data参数 (#356)





 
哪个任务？example里的task脚本、evaluate没怎么优化（batch predict之类的）、个别任务下甚至可能出现耗时比train更长的情况、一个是可以加个tqdm反馈下evaluate的进度、然后嫌慢就自己优化评估的计算过程（主要是一批一批样本处理）、

**Comment by bojone** [comment]
> 问题已解决，需要设置validation_steps=len(valid_generator)，否则会陷入forfit的无限循环中。
> […](#)
> ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: "bojone/bert4keras" ***@***.***&gt;; 发送时间:&nbsp;2021年6月4日(星期五) 下午2:32 ***@***.***&gt;; ***@***.******@***.***&gt;; 主题:&nbsp;Re: [bojone/bert4keras] model.fit时的验证可否直接用validation_data参数 (#356) 哪个任务？example里的task脚本、evaluate没怎么优化（batch predict之类的）、个别任务下甚至可能出现耗时比train更长的情况、一个是可以加个tqdm反馈下evaluate的进度、然后嫌慢就自己优化评估的计算过程（主要是一批一批样本处理）、 — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe.

是的
